api/firewaller/machine.go

```diff
@@ -171,47 +171,63 @@ func (m *Machine) IsManual() (bool, error) {
-func (m *Machine) OpenedMachinePortRanges() (map[names.UnitTag]network.GroupedPortRanges, error) {
+// machine and returns back two maps where keys are unit names and values are
+// the opened port ranges for each unit grouped by endpoint name and subnet
+// CIDR respectively.
+func (m *Machine) OpenedMachinePortRanges() (byEndpoint, byCIDR map[names.UnitTag]network.GroupedPortRanges, err error) {
 	if m.st.BestAPIVersion() < 6 {
 		// OpenedMachinePortRanges() was introduced in FirewallerAPIV6.
-		return nil, errors.NotImplementedf("OpenedMachinePortRanges() (need V6+)")
+		return nil, nil, errors.NotImplementedf("OpenedMachinePortRanges() (need V6+)")
 	}
 
 	var results params.OpenMachinePortRangesResults
 	args := params.Entities{
 		Entities: []params.Entity{{Tag: m.tag.String()}},
 	}
-	err := m.st.facade.FacadeCall("OpenedMachinePortRanges", args, &results)
-	if err != nil {
-		return nil, err
+	if err = m.st.facade.FacadeCall("OpenedMachinePortRanges", args, &results); err != nil {
+		return nil, nil, err
 	}
 	if len(results.Results) != 1 {
-		return nil, fmt.Errorf("expected 1 result, got %d", len(results.Results))
+		return nil, nil, fmt.Errorf("expected 1 result, got %d", len(results.Results))
 	}
 	result := results.Results[0]
 	if result.Error != nil {
-		return nil, result.Error
-	} else if result.GroupKey != "cidr" {
-		return nil, fmt.Errorf("expected open unit port ranges to be grouped by subnet CIDR, got %s", result.GroupKey)
+		return nil, nil, result.Error
+	} else if groupLen := len(result.Groups); groupLen != 2 {
+		return nil, nil, fmt.Errorf("expected two groups for the unit port ranges; got %d", groupLen)
+	} else if result.Groups[0].GroupKey != "endpoint" {
+		return nil, nil, fmt.Errorf("expected first unit port range group to be grouped by endpoint, got %s", result.Groups[0].GroupKey)
+	} else if result.Groups[1].GroupKey != "cidr" {
+		return nil, nil, fmt.Errorf("expected second unit port range group to be grouped by subnet CIDR, got %s", result.Groups[1].GroupKey)
+	}
+
+	if byEndpoint, err = parseGroupedPortRanges(result.Groups[0].UnitPortRanges); err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	if byCIDR, err = parseGroupedPortRanges(result.Groups[1].UnitPortRanges); err != nil {
+		return nil, nil, errors.Trace(err)
 	}
 
-	portRangeMap := make(map[names.UnitTag]network.GroupedPortRanges)
-	for _, unitPortRanges := range result.UnitPortRanges {
+	return byEndpoint, byCIDR, nil
+}
+
+func parseGroupedPortRanges(res []params.OpenUnitPortRanges) (map[names.UnitTag]network.GroupedPortRanges, error) {
+	byUnit := make(map[names.UnitTag]network.GroupedPortRanges)
+	for _, unitPortRanges := range res {
 		unitTag, err := names.ParseUnitTag(unitPortRanges.UnitTag)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		portRangeMap[unitTag] = make(network.GroupedPortRanges)
+		byUnit[unitTag] = make(network.GroupedPortRanges)
 
 		for cidr, portRanges := range unitPortRanges.PortRangeGroups {
```

Comment:
@jameinel Perhaps I should just use a separate payload for the Uniter and Firewaller APIs (they use the same to be consistent with the legacy `OpenedPorts` call that uses the same pattern). This way I can return the grouped by endpoint bit for the uniter call and do something else for the firewaller.

For the firewaller case it might be easier to return a list of {endpoint, CIDR, PortRange} tuples and have the client group however it wants but this would make the response even longer as I would need to repeat the same strings over and over (an endpoint bound to a space with multiple subnets would generate a tuple for each subnet CIDR). What do you think?

Nice spot on the `cidr` bit. It's typo; I will take a look in the morning.

Reply:
Actually, I should be able to group everything down into {endpoint, []CIDR, []PortRange} tuples. I will close this PR and try a different approach tomorrow...